# Human Activity Recognition

# Description

These days Smartphones have become an integral part of our life. We cannot assume our life without a mobile phone. Since, the advent of Smartphones, a revolution has been created in the mobile communication industry. Smartphones are not just restricted for calling these days. Infact, they are more often used for entertainment purpose.

Smartphone manufacturing companies load Smartphones with various sensors to enhance the user experinece. Two of the such sensors are Accelerometer and Gyroscope. Accelerometer measures acceleration while Gyroscope measures angular velocity.

Here, we will try to use the data provided by accelerometer and gyroscope of Smartphone to classify the activity which a Smartphone user is performing

# Why this is Useful?

These days, in addition to Smartphones, we are also using Smart-Watches like Fitbit or Apple-Watch, which help us to track our health. They monitor our each activity throughout the day check how many calories we have burnt. How many hours have we slept. However, in addition to Accelerometer and Gyroscope, they also use Heart-Rate data to monitor our activity. Since, we only have Smartphone data so just by using Accelerometer and Gyroscope data we will monitor the activity of a person. This software can then be converted into an App which can be downloaded in Smartphone. Hence, a person who has Smartphone can monitor his/her health using this App

# Information about Data

# How Data is recorded

The experiments have been carried out with a group of 30 volunteers within an age bracket of 19-48 years. Each person performed six activities (WALKING, WALKING-UPSTAIRS, WALKING-DOWNSTAIRS, SITTING-DOWN, STANDING-UP, LAYING-DOWN) wearing a smartphone (Samsung Galaxy S II) on the waist. Using its embedded accelerometer and gyroscope, we captured 3-axial linear acceleration and 3-axial angular velocity at a constant rate of 50Hz. The experiments have been video-recorded to label the data manually. The obtained dataset has been randomly partitioned into two sets, where 70% of the volunteers was selected for generating the training data and 30% the test data. 5.2. Features

These sensor signals are pre-processed by applying noise filters and then sampled in fixed-width windows(sliding windows) of 2.56 seconds each with 50% overlap. i.e., each window has 128 readings. A 128 size vector is created from each window.
From Each window or to be more precise, from each 128 readings domain experts from signal processing have engineered feature vector of size 561 by calculating variables from the time and frequency domain. In our dataset, each data-point represents a window with different readings.
561 features are stored in the file "Features.docx". Check it out.
Check out 561 features here.(In your blog give here the link of the docx file of features which you upload on github).
The acceleration signal was separated into Body and Gravity acceleration signals(tBodyAcc-XYZ and tGravityAcc-XYZ) using some low pass filter with corner frequency of 0.3Hz.
After that, the body linear acceleration and angular velocity were derived in time to obtain jerk signals (tBodyAccJerk-XYZ and tBodyGyroJerk-XYZ).
The magnitude of these 3-dimensional signals were calculated using the Euclidian norm. This magnitudes are represented as features with names like tBodyAccMag, tGravityAccMag, tBodyAccJerkMag, tBodyGyroMag and tBodyGyroJerkMag.
Finally, We've got frequency domain signals from some of the available signals by applying a FFT (Fast Fourier Transform). These signals obtained were labelled with prefix 'f' just like original signals with prefix 't'. These signals are labelled as fBodyAcc-XYZ, fBodyGyroMag etc.

These are the signals that we got so far.

tBodyAcc-XYZ
tGravityAcc-XYZ
tBodyAccJerk-XYZ
tBodyGyro-XYZ
tBodyGyroJerk-XYZ
tBodyAccMag
tGravityAccMag
tBodyAccJerkMag
tBodyGyroMag
tBodyGyroJerkMag
fBodyAcc-XYZ
fBodyAccJerk-XYZ
fBodyGyro-XYZ
fBodyAccMag
fBodyAccJerkMag
fBodyGyroMag
fBodyGyroJerkMag

9 We can estimate some set of variables from the above signals. i.e., We will estimate the following properties on each and every signal that we recorded so far.

mean(): Mean value
std(): Standard deviation
mad(): Median absolute deviation
max(): Largest value in array
min(): Smallest value in array
sma(): Signal magnitude area
energy(): Energy measure. Sum of the squares divided by the number of values.
iqr(): Inter-quartile range
entropy(): Signal entropy
arCoeff(): Auto-regression coefficients with Burg order equal to 4
correlation(): correlation coefficient between two signals
maxInds(): index of the frequency component with largest magnitude
meanFreq(): Weighted average of the frequency components to obtain a mean frequency
skewness(): skewness of the frequency domain signal
kurtosis(): kurtosis of the frequency domain signal
bandsEnergy(): Energy of a frequency interval within the 64 bins of the FFT of each window.
angle(): Angle between to vectors.


10 We can obtain some other vectors by taking the average of signals in a single window sample. These are used on the angle() variable.

In [19]:
gravityMean
tBodyAccMean
tBodyAccJerkMean
tBodyGyroMean
tBodyGyroJerkMean

NameError: name 'gravityMean' is not defined

# Data Source

Data is downloaded from following source:
https://archive.ics.uci.edu/ml/datasets/human+activity+recognition+using+smartphones

# Quick Overview of Dataset

Accelerometer and Gyroscope readings are taken from 30 volunteers(referred as subjects) while performing the following 6 Activities.

These activites are encoded as follows:
WALKING-- 1
WALKING_UPSTAIRS-- 2
WALKING_DOWNSTAIRS-- 3
SITTING-- 4
STANDING-- 5
LYING-- 6

1.Readings are divided into a window of 2.56 seconds with 50% overlapping.
2.Accelerometer readings are divided into gravity acceleration and body acceleration readings, which has x, y and z components each.
3.Gyroscope readings are the measure of angular velocities which has x, y and z components.
4.Jerk signals are calculated for Body-Acceleration readings.
5.Fourier Transforms are made on the above time readings to obtain frequency readings.
6.Now, on all the base signal readings., mean, max, mad, sma, arcoefficient, energy-bands, entropy etc., are calculated for each window.
7.Extra features are calculated by taking the average of signals in a single window sample. These are used on the angle() variable.
8.Finally, we got feature vector of 561 features and these features are given in the dataset.
Each window of readings is a data-point of 561 features.

# Y-Encoded Labels

In [ ]:
WALKING-- 1
WALKING_UPSTAIRS-- 2
WALKING_DOWNSTAIRS-- 3
SITTING-- 4
STANDING-- 5
LYING-- 6

# Business Problem

Work-flow is as follows:

1.Domain experts from the field of Signal Processing collects the data from Accelerometer and Gyroscope of Smartphone.
2.They break up the data in the time window of 2.56 seconds with 50% overlapping i.e., 128 reading
3.They engineered 561 features from each time window of 2.56 seconds.

By using either human engineered 561 feature data or raw features of 128 reading, our goal is to predict one of the six activities that a Smartphone user is performing at that 2.56 Seconds time window.

# Problem Statement

By using either human engineered 561 feature data or raw features of 128 reading, our goal is to predict one of the six activities that a Smartphone user is performing at that 2.56 Seconds time window.

# Objective and Constraints

1.No Low latency requirement.
2.Errors are not much costly.

All of the Accelerometer and Gyroscope are tri-axial, means that they measure acceleration and angular-velocity respectively in all the three axis namely X-axis, Y-axis and Z-axis. So, we have in total six time-series data. Given this six time-series data, we want to predict six activities namely Walking or Walking-Upstairs or Walking-Downstairs or Lying-Down or Standing-Up or Sitting-Down.

At the outset, this is a multi-class classification problem.

# ML Problem Formulation

All of the Accelerometer and Gyroscope are tri-axial, means that they measure acceleration and angular-velocity respectively in all the three axis namely X-axis, Y-axis and Z-axis. So, we have in total six time-series data. Given this six time-series data, we want to predict six activities namely Walking or Walking-Upstairs or Walking-Downstairs or Lying-Down or Standing-Up or Sitting-Down.

At the outset, this is a multi-class classification problem.

# Performance Metric

1.We will use Accuracy as one of the metric.
2.We will also use Confusion-Matrix to check that in which two activities our model is confused and predicting incorrect activity. For example, between Standing-Up and Sitting-Down. Between Walking-Upstairs and Walking-Downstairs

# Data

All the data is present in 'UCI_HAR_dataset/' folder in present working directory.
Feature names are present in 'UCI_HAR_dataset/features.txt'

Train Data
'UCI_HAR_dataset/train/X_train.txt'
'UCI_HAR_dataset/train/subject_train.txt'
'UCI_HAR_dataset/train/y_train.txt'
Test Data
'UCI_HAR_dataset/test/X_test.txt'
'UCI_HAR_dataset/test/subject_test.txt'
'UCI_HAR_dataset/test/y_test.txt'

# Data-Points Distribution

1.30 test-subjects data is randomly split to 70%(21) train and 30%(7) test data.
2.Each data-point corresponds one of the 6 Activities.

# Plan of Action

We will apply classical Machine Learning models on these 561 sized domain expert engineered features.
As we know that LSTM works well on time-series data, so we have decided that we will apply LSTM of Recurrent Neural Networks on 128 sized raw readings that we obtained from accelerometer and gyroscope signals.

In [ ]:
# Importing Libraries

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Activities are the class labels
# It is a 6 class classification
ACTIVITIES = {
    0: 'WALKING',
    1: 'WALKING_UPSTAIRS',
    2: 'WALKING_DOWNSTAIRS',
    3: 'SITTING',
    4: 'STANDING',
    5: 'LAYING',
}

# Utility function to print the confusion matrix
def confusion_matrix(Y_true, Y_pred):
    Y_true = pd.Series([ACTIVITIES[y] for y in np.argmax(Y_true, axis=1)])
    Y_pred = pd.Series([ACTIVITIES[y] for y in np.argmax(Y_pred, axis=1)])

    return pd.crosstab(Y_true, Y_pred, rownames=['True'], colnames=['Pred'])

### Data

In [3]:
# Data directory
DATADIR = 'UCI_HAR_Dataset'

In [4]:
# Raw data signals
# Signals are from Accelerometer and Gyroscope
# The signals are in x,y,z directions
# Sensor signals are filtered to have only body acceleration
# excluding the acceleration due to gravity
# Triaxial acceleration from the accelerometer is total acceleration
SIGNALS = [
    "body_acc_x",
    "body_acc_y",
    "body_acc_z",
    "body_gyro_x",
    "body_gyro_y",
    "body_gyro_z",
    "total_acc_x",
    "total_acc_y",
    "total_acc_z"
]

In [5]:
# Utility function to read the data from csv file
def _read_csv(filename):
    return pd.read_csv(filename, delim_whitespace=True, header=None)

# Utility function to load the load
def load_signals(subset):
    signals_data = []

    for signal in SIGNALS:
        filename = f'UCI_HAR_Dataset/{subset}/Inertial Signals/{signal}_{subset}.txt'
        signals_data.append(
            _read_csv(filename).as_matrix()
        ) 

    # Transpose is used to change the dimensionality of the output,
    # aggregating the signals by combination of sample/timestep.
    # Resultant shape is (7352 train/2947 test samples, 128 timesteps, 9 signals)
    return np.transpose(signals_data, (1, 2, 0))

In [6]:

def load_y(subset):
    """
    The objective that we are trying to predict is a integer, from 1 to 6,
    that represents a human activity. We return a binary representation of 
    every sample objective as a 6 bits vector using One Hot Encoding
    (https://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html)
    """
    filename = f'UCI_HAR_Dataset/{subset}/y_{subset}.txt'
    y = _read_csv(filename)[0]

    return pd.get_dummies(y).as_matrix()

In [7]:
def load_data():
    """
    Obtain the dataset from multiple files.
    Returns: X_train, X_test, y_train, y_test
    """
    X_train, X_test = load_signals('train'), load_signals('test')
    y_train, y_test = load_y('train'), load_y('test')

    return X_train, X_test, y_train, y_test

In [8]:
# Importing tensorflow
np.random.seed(42)
import tensorflow as tf
tf.set_random_seed(42)

C:\Users\Saurabh\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


In [9]:
# Configuring a session
session_conf = tf.ConfigProto(
    intra_op_parallelism_threads=1,
    inter_op_parallelism_threads=1
)

In [10]:
# Import Keras
from keras import backend as K
sess = tf.Session(graph=tf.get_default_graph(), config=session_conf)
K.set_session(sess)

Using TensorFlow backend.


In [11]:
# Importing libraries
from keras.models import Sequential
from keras.layers import LSTM
from keras.layers.core import Dense, Dropout

In [30]:
# Initializing parameters
epochs = 300
batch_size = 32
n_hidden = 64

In [31]:
# Utility function to count the number of classes
def _count_classes(y):
    return len(set([tuple(category) for category in y]))

In [32]:
# Loading the train and test data
X_train, X_test, Y_train, Y_test = load_data()

C:\Users\Saurabh\Anaconda3\lib\site-packages\ipykernel_launcher.py:12: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  if sys.path[0] == '':


In [33]:
timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
n_classes = _count_classes(Y_train)

print(timesteps)
print(input_dim)
print(len(X_train))

128
9
7352


- Defining the Architecture of LSTM

In [34]:
# Initiliazing the sequential model
model = Sequential()
# Configuring the parameters
model.add(LSTM(n_hidden, input_shape=(timesteps, input_dim)))
# Adding a dropout layer
model.add(Dropout(0.5))
# Adding a dense output layer with sigmoid activation
model.add(Dense(n_classes, activation='sigmoid'))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
lstm_3 (LSTM)                (None, 64)                18944     
_________________________________________________________________
dropout_3 (Dropout)          (None, 64)                0         
_________________________________________________________________
dense_3 (Dense)              (None, 6)                 390       
Total params: 19,334
Trainable params: 19,334
Non-trainable params: 0
_________________________________________________________________


In [35]:
# Compiling the model
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

In [36]:
# Training the model
model.fit(X_train,
          Y_train,
          batch_size=batch_size,
          validation_data=(X_test, Y_test),
          epochs=epochs)

Train on 7352 samples, validate on 2947 samples
Epoch 1/300
7352/7352 [==============================] - 24s 3ms/step - loss: 1.4284 - acc: 0.3644 - val_loss: 1.3755 - val_acc: 0.3431
Epoch 2/300
7352/7352 [==============================] - 22s 3ms/step - loss: 1.3148 - acc: 0.3923 - val_loss: 1.3391 - val_acc: 0.4262
Epoch 3/300
7352/7352 [==============================] - 24s 3ms/step - loss: 1.2377 - acc: 0.4463 - val_loss: 1.2617 - val_acc: 0.4740s - loss: 1.269
Epoch 4/300
7352/7352 [==============================] - 23s 3ms/step - loss: 1.1586 - acc: 0.4737 - val_loss: 1.3795 - val_acc: 0.4092
Epoch 5/300
7352/7352 [==============================] - 22s 3ms/step - loss: 1.2517 - acc: 0.4286 - val_loss: 1.2629 - val_acc: 0.3997
Epoch 6/300
7352/7352 [==============================] - 22s 3ms/step - loss: 1.1608 - acc: 0.5076 - val_loss: 1.0618 - val_acc: 0.5643
Epoch 7/300
7352/7352 [==============================] - 22s 3ms/step - loss: 0.9851 - acc: 0.5613 - val_loss: 0.9977 - v

7352/7352 [==============================] - 22s 3ms/step - loss: 0.1785 - acc: 0.9389 - val_loss: 0.3258 - val_acc: 0.8935
Epoch 60/300
7352/7352 [==============================] - 22s 3ms/step - loss: 0.1423 - acc: 0.9491 - val_loss: 0.3077 - val_acc: 0.9006
Epoch 61/300
7352/7352 [==============================] - 22s 3ms/step - loss: 0.1715 - acc: 0.9368 - val_loss: 0.3258 - val_acc: 0.9077
Epoch 62/300
7352/7352 [==============================] - 22s 3ms/step - loss: 0.1732 - acc: 0.9410 - val_loss: 0.2511 - val_acc: 0.9104
Epoch 63/300
7352/7352 [==============================] - 22s 3ms/step - loss: 0.1449 - acc: 0.9468 - val_loss: 0.2585 - val_acc: 0.9125
Epoch 64/300
7352/7352 [==============================] - 22s 3ms/step - loss: 0.1437 - acc: 0.9476 - val_loss: 0.2573 - val_acc: 0.9135
Epoch 65/300
7352/7352 [==============================] - 22s 3ms/step - loss: 0.1262 - acc: 0.9518 - val_loss: 0.2454 - val_acc: 0.9169
Epoch 66/300
7352/7352 [==============================

7352/7352 [==============================] - 22s 3ms/step - loss: 0.1476 - acc: 0.9427 - val_loss: 0.3411 - val_acc: 0.9097
Epoch 119/300
7352/7352 [==============================] - 22s 3ms/step - loss: 0.1338 - acc: 0.9421 - val_loss: 0.2694 - val_acc: 0.9158
Epoch 120/300
7352/7352 [==============================] - 22s 3ms/step - loss: 0.1197 - acc: 0.9505 - val_loss: 0.2509 - val_acc: 0.9277
Epoch 121/300
7352/7352 [==============================] - 22s 3ms/step - loss: 0.1667 - acc: 0.9380 - val_loss: 0.3682 - val_acc: 0.9128
Epoch 122/300
7352/7352 [==============================] - 22s 3ms/step - loss: 0.1307 - acc: 0.9463 - val_loss: 0.2353 - val_acc: 0.9260
Epoch 123/300
7352/7352 [==============================] - 22s 3ms/step - loss: 0.1149 - acc: 0.9516 - val_loss: 0.2196 - val_acc: 0.9250
Epoch 124/300
7352/7352 [==============================] - 22s 3ms/step - loss: 0.1127 - acc: 0.9512 - val_loss: 0.2116 - val_acc: 0.9308
Epoch 125/300
7352/7352 [=======================

7352/7352 [==============================] - 38s 5ms/step - loss: 0.1093 - acc: 0.9557 - val_loss: 0.2953 - val_acc: 0.9199
Epoch 178/300
7352/7352 [==============================] - 38s 5ms/step - loss: 0.1101 - acc: 0.9555 - val_loss: 0.2767 - val_acc: 0.9257
Epoch 179/300
7352/7352 [==============================] - 38s 5ms/step - loss: 0.1004 - acc: 0.9573 - val_loss: 0.3647 - val_acc: 0.9135
Epoch 180/300
7352/7352 [==============================] - 38s 5ms/step - loss: 0.1120 - acc: 0.9547 - val_loss: 0.3763 - val_acc: 0.9128
Epoch 181/300
7352/7352 [==============================] - 34s 5ms/step - loss: 0.1104 - acc: 0.9510 - val_loss: 0.3103 - val_acc: 0.9220
Epoch 182/300
7352/7352 [==============================] - 33s 5ms/step - loss: 0.1223 - acc: 0.9531 - val_loss: 0.2781 - val_acc: 0.9220
Epoch 183/300
7352/7352 [==============================] - 33s 4ms/step - loss: 0.1390 - acc: 0.9354 - val_loss: 0.3215 - val_acc: 0.9040
Epoch 184/300
7352/7352 [=======================

7352/7352 [==============================] - 40s 5ms/step - loss: 0.1107 - acc: 0.9567 - val_loss: 0.2582 - val_acc: 0.9209
Epoch 237/300
7352/7352 [==============================] - 40s 5ms/step - loss: 0.1041 - acc: 0.9589 - val_loss: 0.2747 - val_acc: 0.9213
Epoch 238/300
7352/7352 [==============================] - 40s 5ms/step - loss: 0.1127 - acc: 0.9548 - val_loss: 0.3241 - val_acc: 0.9260
Epoch 239/300
7352/7352 [==============================] - 40s 5ms/step - loss: 0.1108 - acc: 0.9577 - val_loss: 0.3341 - val_acc: 0.9141
Epoch 240/300
7352/7352 [==============================] - 40s 5ms/step - loss: 0.1044 - acc: 0.9584 - val_loss: 0.3490 - val_acc: 0.9216
Epoch 241/300
7352/7352 [==============================] - 39s 5ms/step - loss: 0.1019 - acc: 0.9593 - val_loss: 0.3596 - val_acc: 0.9179
Epoch 242/300
7352/7352 [==============================] - 40s 5ms/step - loss: 0.0995 - acc: 0.9580 - val_loss: 0.3850 - val_acc: 0.9158
Epoch 243/300
7352/7352 [=======================

7352/7352 [==============================] - 34s 5ms/step - loss: 0.1000 - acc: 0.9573 - val_loss: 0.3639 - val_acc: 0.9352
Epoch 296/300
7352/7352 [==============================] - 32s 4ms/step - loss: 0.1204 - acc: 0.9567 - val_loss: 0.2257 - val_acc: 0.9413
Epoch 297/300
7352/7352 [==============================] - 33s 4ms/step - loss: 0.0881 - acc: 0.9604 - val_loss: 0.2184 - val_acc: 0.9389
Epoch 298/300
7352/7352 [==============================] - 34s 5ms/step - loss: 0.0848 - acc: 0.9618 - val_loss: 0.2433 - val_acc: 0.9393
Epoch 299/300
7352/7352 [==============================] - 33s 5ms/step - loss: 0.0939 - acc: 0.9603 - val_loss: 0.1992 - val_acc: 0.9355
Epoch 300/300
7352/7352 [==============================] - 32s 4ms/step - loss: 0.0990 - acc: 0.9582 - val_loss: 0.2177 - val_acc: 0.9325


In [37]:
# Confusion Matrix
print(confusion_matrix(Y_test, model.predict(X_test)))

Pred                LAYING  SITTING  STANDING  WALKING  WALKING_DOWNSTAIRS  \
True                                                                         
LAYING                 537        0         0        0                   0   
SITTING                  0      378       113        0                   0   
STANDING                 0       59       473        0                   0   
WALKING                  0        0         0      496                   0   
WALKING_DOWNSTAIRS       0        1         0       12                 399   
WALKING_UPSTAIRS         0        2         1        3                   0   

Pred                WALKING_UPSTAIRS  
True                                  
LAYING                             0  
SITTING                            0  
STANDING                           0  
WALKING                            0  
WALKING_DOWNSTAIRS                 8  
WALKING_UPSTAIRS                 465  


In [38]:
score = model.evaluate(X_test, Y_test)

2947/2947 [==============================] - 3s 897us/step


In [39]:
score

[0.21770981482614904, 0.9324737020699015]

- With a simple 2 layer architecture we got 93.24% accuracy and a loss of 0.21
- We can further imporve the performace with Hyperparameter tuning

# Conclusion : 

1.Here we seen thatby using two layered LSTM, we got a accuracy of 93.24% with loss of 0.21% .DeeP Learning help us to built models even when we don't have domain expert engineered features.
2.LSTM model can  give good result by doing more hypertuning.
3.In this case study for better accuracy I had checked different epochs with different parameters.